kvbc/src/v4blockchain/detail/st_chain.cpp

```diff
@@ -23,6 +24,76 @@ StChain::StChain(const std::shared_ptr<concord::storage::rocksdb::NativeClient>&
     LOG_INFO(V4_BLOCK_LOG,
              "Created [" << v4blockchain::detail::ST_CHAIN_CF << "] column family for the state transfer blockchain");
   }
+  loadLastBlockId();
+  if (last_block_id_ > 0) {
+    LOG_INFO(CAT_BLOCK_LOG, "State transfer last block id: " << last_block_id_);
+  }
+}
+
+/////////// BLOCKS/////////////////////////
+bool StChain::hasBlock(BlockId block_id) const {
+  if (last_block_id_ < block_id) return false;
+  return native_client_
+      ->getSlice(v4blockchain::detail::ST_CHAIN_CF, v4blockchain::detail::Blockchain::generateKey(block_id))
+      .has_value();
+}
+
+void StChain::addBlock(const BlockId id, const char* block, const uint32_t blockSize) {
+  auto write_batch = native_client_->getBatch();
+  write_batch.put(v4blockchain::detail::ST_CHAIN_CF,
+                  v4blockchain::detail::Blockchain::generateKey(id),
+                  ::rocksdb::Slice(block, blockSize));
+  native_client_->write(std::move(write_batch));
+  updateLastIdIfBigger(id);
+}
+
+void StChain::deleteBlock(const BlockId id, storage::rocksdb::NativeWriteBatch& wb) {
+  wb.del(v4blockchain::detail::ST_CHAIN_CF, v4blockchain::detail::Blockchain::generateKey(id));
+}
+/////////////// ST LAST BLOCK ID ////////////
+void StChain::loadLastBlockId() {
+  auto itr = native_client_->getIterator(v4blockchain::detail::ST_CHAIN_CF);
+  auto max_db_key = v4blockchain::detail::Blockchain::generateKey(v4blockchain::detail::Blockchain::MAX_BLOCK_ID);
+  itr.seekAtMost(max_db_key);
+  if (!itr) {
+    last_block_id_ = 0;
+    return;
+  }
+  last_block_id_ = concordUtils::fromBigEndianBuffer<BlockId>(itr.keyView().data());
+}
+
+void StChain::updateLastIdAfterDeletion(const BlockId id) {
+  if (last_block_id_ == 0 || last_block_id_ != id) {
+    return;
+  }
+  last_block_id_ = 0;
+  loadLastBlockId();
+  return;
+}
+
+void StChain::updateLastIdIfBigger(const BlockId id) {
+  if (last_block_id_ >= id) {
+    return;
+  }
+  last_block_id_ = id;
+}
+
+std::optional<v4blockchain::detail::Block> StChain::getBlock(kvbc::BlockId id) const {
+  auto key = v4blockchain::detail::Blockchain::generateKey(id);
+  auto opt_block_str = native_client_->get(v4blockchain::detail::ST_CHAIN_CF, key);
+  if (!opt_block_str) return std::nullopt;
+  return v4blockchain::detail::Block(*opt_block_str);
+}
+
+concord::util::digest::BlockDigest StChain::getBlockParentDigest(concord::kvbc::BlockId id) const {
+  auto block_str = getBlock(id);
+  ConcordAssert(block_str.has_value());
+  return v4blockchain::detail::Block{*block_str}.parentDigest();
```

Comment:
the get block of the ST chain returns a "raw" block i.e. a string and we should construct a block from it.
but it can be optimized not to construct a block but to access the digest directly from the underlying buffer.

Reply:
you're right, I'll change that